File: src/main/java/hu/gerviba/webschop/web/AdminController.java
package hu.gerviba.webschop.web;

import java.io.IOException;
import java.security.NoSuchAlgorithmException;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

import javax.servlet.http.HttpServletRequest;
import javax.validation.Valid;

import hu.gerviba.webschop.service.ItemPrecedenceService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.stereotype.Controller;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import hu.gerviba.webschop.dao.RoleEntryDto;
import hu.gerviba.webschop.model.CircleEntity;
import hu.gerviba.webschop.model.UserEntity;
import hu.gerviba.webschop.service.CircleService;
import hu.gerviba.webschop.service.ItemService;
import hu.gerviba.webschop.service.UserService;
import io.swagger.annotations.Api;

@Controller
@PreAuthorize("hasRole('ADMIN')")
@RequestMapping("/admin")
@Api(value="onlinestore", description="Admin pages")
public class AdminController {

    @Autowired
    CircleService circles;

    @Autowired
    ItemService items;

    @Autowired
    UserService users;
    
    @Autowired
    ControllerUtil util;

    @Autowired
    ItemPrecedenceService itemSorter;
    
	@GetMapping("/")
	public String adminRoot(Map<String, Object> model) {
        model.put("circles", circles.findAllForMenu());
        model.put("circlesToEdit", circles.findAll());
        List<RoleEntryDto> roles = users.findAll().stream()
                .map(user -> {
                    try {
                        return new RoleEntryDto(util.sha256(user.getUid()), user);
                    } catch (NoSuchAlgorithmException e) {
                        e.printStackTrace();
                    }
                    return null;
                })
                .filter(Objects::nonNull)
                .sorted(Comparator.comparing(RoleEntryDto::getName))
                .collect(Collectors.toList());
        model.put("roles", roles);
		return "admin";
	}

	@GetMapping("/circles")
	public String adminCircles(Map<String, Object> model) {
		model.put("circles", circles.findAllForMenu());
        model.put("circlesToEdit", circles.findAll());
		return "admin";
	}

    @GetMapping("/circles/new")
    public String adminAddCircle(Map<String, Object> model) {
        model.put("circles", circles.findAllForMenu());
        model.put("mode", "new");
        model.put("adminMode", true);
        model.put("circle", new CircleEntity());
        return "circleModify";
    }

    @PostMapping("/circles/new")
    public String adminAddCircle(HttpServletRequest request, 
            @Valid CircleEntity circle, 
            @RequestParam("logo") MultipartFile logo, 
            @RequestParam("background") MultipartFile background) throws IllegalStateException, IOException {
        
        circle.setRateingCount(0);
        circle.setRateOverAll(0);
        circle.setRatePrice(0);
        circle.setRateQuality(0);
        circle.setRateSpeed(0);
        
        String file = util.uploadFile("logos", logo);
        circle.setLogoUrl(file == null ? "image/blank-logo.svg" : "cdn/logos/" + file);
        
        file = util.uploadFile("backgrounds", background);
        circle.setBackgroundUrl(file == null ? "image/blank-background.jpg" : "cdn/backgrounds/" + file);
        
        circles.save(circle);
        
        return "redirect:/admin/";
    }
    
    @GetMapping("/circles/edit/{circleId}")
    public String adminEditCircle(@PathVariable Long circleId, Map<String, Object> model) {
        model.put("circles", circles.findAllForMenu());
        model.put("mode", "edit");
        model.put("adminMode", true);
        model.put("circle", new CircleEntity(circles.getOne(circleId)));
        return "circleModify";
    }

    @PostMapping("/circles/edit")
    public String adminEditCircle(@ModelAttribute @Valid CircleEntity circle, 
    		BindingResult bindingResult,
            @RequestParam Long circleId,
            @RequestParam(required = false) MultipartFile logo,
            @RequestParam(required = false) MultipartFile background,
            Map<String, Object> model) {
        
        if (bindingResult.hasErrors()) {
            model.put("circles", circles.findAllForMenu());
            model.put("mode", "edit");
            model.put("adminMode", true);
            model.put("circle", circle);
            return "circleModify";
        }
        
        CircleEntity original = circles.getOne(circleId);
        
        original.setAvgOpening(circle.getAvgOpening());
        original.setCssClassName(circle.getCssClassName());
        original.setDescription(circle.getDescription());
        original.setDisplayName(circle.getDisplayName());
        original.setFacebookUrl(circle.getFacebookUrl());
        original.setFounded(circle.getFounded());
        original.setHomePageDescription(circle.getHomePageDescription());
        original.setHomePageOrder(circle.getHomePageOrder());
        original.setWebsiteUrl(circle.getWebsiteUrl());
        original.setAlias(circle.getAlias());
        original.setVisible(circle.getVisible());
        
        String file = util.uploadFile("logos", logo);
        if (file != null)
            original.setLogoUrl("cdn/logos/" + file);
        
        file = util.uploadFile("backgrounds", background);
        if (file != null)
            original.setBackgroundUrl("cdn/backgrounds/" + file);
        
        circles.save(original);
        
        return "redirect:/admin/";
    }
    
    @GetMapping("/circles/delete/{circleId}")
    public String adminDeleteCircle(@PathVariable Long circleId, Map<String, Object> model) {
        model.put("circles", circles.findAllForMenu());
        model.put("topic", "circle");
        model.put("arg", circles.getOne(circleId).getDisplayName());
        model.put("ok", "admin/circles/delete/" + circleId + "/confirm");
        model.put("cancel", "admin/");
        return "prompt";
    }

    @PostMapping("/circles/delete/{circleId}/confirm")
    public String adminDeleteCircleConfirm(@PathVariable Long circleId, 
            HttpServletRequest request) {
        
        items.deleteByCircle(circleId);
        circles.delete(circles.getOne(circleId));
        return "redirect:/admin/";
    }
    
    @GetMapping("/roles/edit/{uidHash}")
    public String editRoles(@PathVariable String uidHash, Map<String, Object> model) {
        model.put("circles", circles.findAllForMenu());
        model.put("uidHash", uidHash);
        UserEntity user = users.getByUidHash(uidHash);
        model.put("name", user.getName());
        model.put("sysadmin", user.isSysadmin());
        model.put("roles", String.join(", ", user.getPermissions()));
        model.put("priority", user.getOrderingPriority());
        return "userModify";
    }

    @PostMapping("/roles/edit/")
    public String editRoles(@RequestParam String uidHash,
            @RequestParam String roles,
            @RequestParam(required = false, defaultValue = "false") boolean sysadmin,
            @RequestParam(defaultValue = "1") Integer priority
    ) {
        
        UserEntity user = users.getByUidHash(uidHash);
        user.setSysadmin(sysadmin);
        user.getPermissions().clear();
        for (String role : roles.split(",")) {
            role = role.trim().toUpperCase();
            if (role.startsWith("CIRCLE_"))
                user.getPermissions().add(role);
            if (role.startsWith("PR_"))
                user.getPermissions().add(role);
        }
        if (user.getPermissions().size() > 0)
            user.getPermissions().add("ROLE_LEADER");

        user.setOrderingPriority(priority);
        users.save(user);

        return "redirect:/admin/";
    }

    @ResponseBody
    @GetMapping("/circles/api/reorder")
    public String reorder() {
        itemSorter.reorder();
        return "ok";
    }

}

File: src/main/kotlin/hu/kirdev/schpincer/web/AdminController.kt
package hu.kirdev.schpincer.web

import hu.kirdev.schpincer.config.ApplicationConfig
import hu.kirdev.schpincer.config.Role
import hu.kirdev.schpincer.dao.UserRepository
import hu.kirdev.schpincer.dto.RoleEntryDto
import hu.kirdev.schpincer.model.CardType
import hu.kirdev.schpincer.model.CircleEntity
import hu.kirdev.schpincer.service.*
import jakarta.servlet.http.HttpServletRequest
import jakarta.validation.Valid
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.validation.BindingResult
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile
import java.io.IOException
import java.util.*

const val REDIRECT_TO_ADMIN = "redirect:/admin"


@Controller
@RequestMapping("/admin")
open class AdminController(
    private val circles: CircleService,
    private val items: ItemService,
    private val users: UserService,
    private val itemSorter: ItemPrecedenceService,
    private val config: RealtimeConfigService,
    private val userRepository: UserRepository,
    private val app: ApplicationConfig
) {

    @GetMapping("")
    fun adminRoot(model: Model): String {
        model.addAttribute("circles", circles.findAllForMenu())
        model.addAttribute("circlesToEdit", circles.findAll())
        val roles: List<RoleEntryDto> = users.findAll()
            .map { RoleEntryDto(it.uid, it) }
            .sortedBy { it.name }
        model.addAttribute("roles", roles)
        model.addAttribute("configObject", ConfigObject(config))
        config.injectPublicValues(model)
        return "admin"
    }

    @GetMapping("/circles")
    fun adminCircles(model: Model): String {
        model.addAttribute("circles", circles.findAllForMenu())
        model.addAttribute("circlesToEdit", circles.findAll())
        config.injectPublicValues(model)
        return "admin"
    }

    @GetMapping("/circles/new")
    fun adminAddCircle(model: Model): String {
        model.addAttribute("circles", circles.findAllForMenu())
        model.addAttribute("mode", "new")
        model.addAttribute("adminMode", true)
        model.addAttribute("circle", CircleEntity())
        config.injectPublicValues(model)
        return "circleModify"
    }

    @PostMapping("/circles/new")
    @Throws(IllegalStateException::class, IOException::class)
    fun adminAddCircle(
        request: HttpServletRequest,
        circle: @Valid CircleEntity,
        @RequestParam("logo") logo: MultipartFile?,
        @RequestParam("background") background: MultipartFile?
    ): String {

        var file = logo?.uploadFile(app.uploadPath, "logos")
        circle.logoUrl = if (file == null) "image/blank-logo.svg" else "cdn/logos/$file"

        file = background?.uploadFile(app.uploadPath, "backgrounds")
        circle.backgroundUrl = if (file == null) "image/blank-background.jpg" else "cdn/backgrounds/$file"

        circles.save(circle)
        return REDIRECT_TO_ADMIN
    }

    @GetMapping("/circles/edit/{circleId}")
    fun adminEditCircle(@PathVariable circleId: Long, model: Model): String {
        model.addAttribute("circles", circles.findAllForMenu())
        model.addAttribute("mode", "edit")
        model.addAttribute("adminMode", true)
        model.addAttribute("circle", circles.getOne(circleId)?.copy() ?: return "redirect:/admin/?error=invalidCircleId")
        config.injectPublicValues(model)
        return "circleModify"
    }

    @PostMapping("/circles/edit")
    fun adminEditCircle(
        @ModelAttribute circle:
        @Valid CircleEntity,
        bindingResult: BindingResult,
        @RequestParam circleId: Long,
        @RequestParam(required = false) logo: MultipartFile?,
        @RequestParam(required = false) background: MultipartFile?,
        model: Model
    ): String {

        if (bindingResult.hasErrors()) {
            model.addAttribute("circles", circles.findAllForMenu())
            model.addAttribute("mode", "edit")
            model.addAttribute("adminMode", true)
            model.addAttribute("circle", circle)
            return "circleModify"
        }
        val original = circles.getOne(circleId) ?: return "redirect:/admin/?error=invalidCircleId"

        with(original) {
            avgOpening = circle.avgOpening
            cssClassName = circle.cssClassName
            description = circle.description
            displayName = circle.displayName
            facebookUrl = circle.facebookUrl
            founded = circle.founded
            homePageDescription = circle.homePageDescription
            homePageOrder = circle.homePageOrder
            websiteUrl = circle.websiteUrl
            alias = circle.alias
            visible = circle.visible
            virGroupId = circle.virGroupId

            var file = logo?.uploadFile(app.uploadPath, "logos")
            if (file != null)
                logoUrl = "cdn/logos/$file"

            file = background?.uploadFile(app.uploadPath, "backgrounds")
            if (file != null)
                backgroundUrl = "cdn/backgrounds/$file"
        }
        circles.save(original)
        return REDIRECT_TO_ADMIN
    }

    @GetMapping("/circles/delete/{circleId}")
    fun adminDeleteCircle(@PathVariable circleId: Long, model: Model): String {
        model.addAttribute("circles", circles.findAllForMenu())
        model.addAttribute("topic", "circle")
        model.addAttribute("arg", circles.getOne(circleId)?.displayName ?: return "redirect:/admin/?error=invalidCircleId")
        model.addAttribute("ok", "admin/circles/delete/$circleId/confirm")
        model.addAttribute("cancel", "admin/")
        config.injectPublicValues(model)
        return "prompt"
    }

    @PostMapping("/circles/delete/{circleId}/confirm")
    fun adminDeleteCircleConfirm(
        @PathVariable circleId: Long,
        request: HttpServletRequest
    ): String {
        items.deleteByCircle(circleId)
        val circle = circles.getOne(circleId) ?: return "redirect:/admin/?error=invalidCircleId"
        circles.delete(circle)
        return REDIRECT_TO_ADMIN
    }

    @GetMapping("/roles/edit/{uid}")
    fun editRoles(@PathVariable uid: String, model: Model): String {
        model.addAttribute("circles", circles.findAllForMenu())
        model.addAttribute("uid", uid)
        val user = users.getByIdOrNull(uid) ?: return "redirect:/admin/?error=invalidUid"
        model.addAttribute("name", user.name)
        model.addAttribute("sysadmin", user.sysadmin)
        model.addAttribute("email", user.email ?: "-")
        model.addAttribute("roles", user.permissions.joinToString(", "))
        model.addAttribute("priority", user.orderingPriority)
        model.addAttribute("forceGrantLoginAccess", user.forceGrantLoginAccess)
        model.addAttribute("alwaysGrantAb", user.alwaysGrantAb)
        config.injectPublicValues(model)
        return "userModify"
    }

    @PostMapping("/roles/edit")
    fun editRoles(
        @RequestParam uid: String,
        @RequestParam roles: String,
        @RequestParam(required = false, defaultValue = "false") sysadmin: Boolean,
        @RequestParam(defaultValue = "1") priority: Int,
        @RequestParam(required = false, defaultValue = "false") forceGrantLoginAccess: Boolean,
        @RequestParam(required = false, defaultValue = "false") alwaysGrantAb: Boolean,
    ): String {
        val user = users.getByIdOrNull(uid) ?: return "redirect:/admin/?error=invalidUid"
        user.sysadmin = sysadmin
        val permissions = mutableSetOf<String>()
        for (role in roles.split(",".toRegex()).toTypedArray()) {
            val roleName = role.trim().uppercase(Locale.getDefault())
            if (roleName.startsWith("CIRCLE_"))
                permissions.add(roleName)
            if (roleName.startsWith("PR_"))
                permissions.add(roleName)
        }
        if (permissions.isNotEmpty())
            permissions.add("ROLE_${Role.LEADER.name}")
        user.permissions = permissions
        user.orderingPriority = priority
        user.forceGrantLoginAccess = forceGrantLoginAccess
        user.alwaysGrantAb = alwaysGrantAb
        users.save(user)
        return REDIRECT_TO_ADMIN
    }

    @ResponseBody
    @GetMapping("/circles/api/reorder")
    fun reorder(): String {
        itemSorter.reorder()
        return "ok"
    }

    @ResponseBody
    @GetMapping("/debug/card/{card}")
    fun changeCard(@PathVariable card: String, auth: Authentication?): String {
        val user = auth.getUser(userRepository) ?: return "failed"
        val cardType = try {
            CardType.valueOf(card)
        } catch (e: IllegalArgumentException) {
            return "failed"
        }
        user.pekCardType = cardType
        users.save(user)
        return "ok"
    }

    @PostMapping("/config")
    fun adminConfigUpdate(@ModelAttribute configObject: ConfigObject): String {
        config.messageBoxType = configObject.messageBoxType
        config.messageBoxMessage = configObject.messageBoxMessage
        return REDIRECT_TO_ADMIN
    }

}
